tools/convert_bigfile_gadget_format3.py

```diff
@@ -0,0 +1,271 @@
+    battr["RSDFactor"] = np.nan
+    #Pass other keys through unchanged. We whitelist expected keys to avoid confusing Gadget.
+    hdfats = ["MassTable", "Time", "BoxSize", "Omega0", "OmegaLambda", "HubbleParam"]
+    for attr in hdfats:
+        battr[attr] = hattr[attr]
+
+def write_bf_segment(bf, hfile, startpart):
+    """Write the data arrays to an HDF5 file."""
+    #Open the file
+    with h5py.File(hfile,'r') as hdf5:
+        endpart = startpart + hdf5["Header"].attrs["NumPart_ThisFile"]
+        for ptype in range(6):
+            try:
+                htype = hdf5["PartType"+str(ptype)]
+            except KeyError:
+                continue
+            for hname in htype.keys():
+                block = names.get_bigfile_name(hname)
+                bname = "%d/%s" % (ptype, block)
+                harray = np.array(hdf5["PartType"+str(ptype)][hname])
+                #Beware this is not checked.
+                bf[bname].write(startpart[ptype], harray)
+        return endpart
+
+def create_big_file_arrays(bf, hfile):
+    """Pre-create all the big file arrays with the desired sizes."""
+    npart_tot = bf["Header"].attrs["TotNumPart"]
+    nfile = hfile["Header"].attrs["NumFilesPerSnapshot"]
+    for ptype in range(6):
+        try:
+            htype = hfile["PartType"+str(ptype)]
+        except KeyError:
+            continue
+        for hname in htype.keys():
+            hshape = np.shape(htype[hname])
+            dtype = htype[hname].dtype
+            block = names.get_bigfile_name(hname)
+            bname = "%d/%s" % (ptype, block)
+            try:
+                rows = hshape[1]
+            except IndexError:
+                rows = 1
+            bf.create(bname, dtype=(dtype, rows), size=npart_tot[ptype], Nfile=nfile)
+        #We need these (generally non-present) arrays but the simulation will start even if they are zero.
+        if ptype == 0 and "Generation" not in htype.keys():
+            genzero = np.zeros(npart_tot[ptype],dtype=np.uint8)
+            bf.create_from_array("0/Generation", genzero)
+        #MP-Gadget needs a mass array: Gadget-3 reads it from the header if all masses are the same.
+        if "Mass" not in htype.keys():
+            genzero = hfile["Header"].attrs["MassTable"][ptype] * np.ones(npart_tot[ptype],dtype=np.float32)
+            bf.create_from_array("%d/Mass" % ptype, genzero)
+
+
+def write_big_file(bfname, hdf5name):
+    """Find all the HDF5 files in the snapshot and merge them into a bigfile."""
+    #Find all the HDF5 snapshot set.
+    hdf5_files = glob.glob(hdf5name)
+    if len(hdf5_files) == 0:
+        hdf5_files = glob.glob(hdf5name+".*.hdf5")
+    elif os.path.isdir(hdf5_files[0]):
+        hdf5_files = glob.glob(os.path.join(hdf5name,"*_[0-9][0-9][0-9].*.hdf5"))
+    if len(hdf5_files) == 0:
+        raise IOError("Could not find hdF5 snapshot as %s (.*.hdf5)" % hdf5name)
+    #Sort so we get a consistent answer each time.
+    hdf5_files.sort()
```

Comment:
glob.glob returns a list or a generator? may use sorted(....)

Reply:
We could, is that preferred?

Reply:
I thought they did the same thing, just one is in place and one is not?

Reply:
Yes. functional better than statement, when we can afford.

Reply:
So you prefer hdf5_files = sorted(hdf5_files)? Really? Just because glob.glob might return a generator? I can change it but I feel that this egg should be broken at the other end! :-)

Reply:
sorted builtin function was added in Python 2.4, and it appears it is preferred than list.sort in most of the examples in https://docs.python.org/3/howto/sorting.html#sortinghowto (which could not be done with sort). 

The main difference is sorted is a function, while .sort() behaves like a statement. A function is clearer than a statement because a statement only performs side-effects, like modifying the argument in-place.

Reply:
Ok, changed.